File: MPP/MPP Proiect/Java/Networking/src/main/java/protobuffprotocol/ProtoProxy.java
package protobuffprotocol;

import model.Participant;
import model.Referee;
import model.Test;
import service.IObserver;
import service.IService;
import service.MyException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class ProtoProxy implements IService{
    private final String host;
    private final int port;

    private IObserver client;

    private InputStream input;
    private OutputStream output;
    private Socket connection;

    private final BlockingQueue<Protobufs.Response> qresponses;
    private volatile boolean finished;

    public ProtoProxy(String host, int port){
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingQueue<>();
    }


    @Override
    public Referee login(Referee referee, IObserver refereeObserver) throws MyException{
        initializeConnection();
        System.out.println("Login request ...");
        sendRequest(ProtoUtils.createLoginRequest(referee));
        Protobufs.Response response = readResponse();
        if(response.getType() == Protobufs.Response.Type.OK){
            this.client = refereeObserver;
        }
        if(response.getType() == Protobufs.Response.Type.ERROR){
            String errorText = ProtoUtils.getError(response);
            closeConnection();
            throw new MyException(errorText);
        }
        return ProtoUtils.getReferee(response);
    }

    @Override
    public Iterable<Participant> getAllParticipants() throws MyException{
        sendRequest(ProtoUtils.createGetAllParticipantsRequest());
        Protobufs.Response response = readResponse();
        if(response.getType() == Protobufs.Response.Type.ERROR){
            String errorText = ProtoUtils.getError(response);
            throw new MyException(errorText);
        }
        return ProtoUtils.getAllParticipants(response);
    }

    @Override
    public void addTest(Test test) throws MyException{
        sendRequest(ProtoUtils.createAddTestRequest(test));
        Protobufs.Response response = readResponse();
        if(response.getType() == Protobufs.Response.Type.ERROR){
            String errorText = ProtoUtils.getError(response);
            throw new MyException(errorText);
        }
    }

    @Override
    public Iterable<Participant> getAllActiveParticipants(int id) throws MyException{
        sendRequest(ProtoUtils.createGetAllActiveParticipantsRequest(id));
        Protobufs.Response response = readResponse();
        if(response.getType() == Protobufs.Response.Type.ERROR){
            String errorText = ProtoUtils.getError(response);
            throw new MyException(errorText);
        }
        return ProtoUtils.getAllParticipants(response);
    }

    @Override
    public void logout(Referee referee, IObserver refereeObserver) throws MyException{
        sendRequest(ProtoUtils.createLogoutRequest(referee));
        Protobufs.Response response = readResponse();
        closeConnection();
        if(response.getType() == Protobufs.Response.Type.ERROR){
            String errorText = ProtoUtils.getError(response);
            throw new MyException(errorText);
        }
    }

    private void closeConnection(){
        finished = true;
        try{
            input.close();
            output.close();
            connection.close();
            client = null;
        }catch(IOException e){
            e.printStackTrace();
        }
    }

    private void sendRequest(Protobufs.Request request) throws MyException{
        try{
            System.out.println("Sending request ..." + request);
            //request.writeTo(output);
            request.writeDelimitedTo(output);
            output.flush();
            System.out.println("Request sent.");
        }catch(IOException e){
            throw new MyException("Error sending object " + e);
        }

    }

    private Protobufs.Response readResponse() throws MyException{
        Protobufs.Response response = null;
        try{
            response = qresponses.take();

        }catch(InterruptedException e){
            e.printStackTrace();
        }
        return response;
    }

    private void initializeConnection() throws MyException{
        try{
            connection = new Socket(host, port);
            output = connection.getOutputStream();
            //output.flush();
            input = connection.getInputStream();
            //new ObjectInputStream(connection.getInputStream());
            finished = false;
            startReader();
        }catch(IOException e){
            e.printStackTrace();
        }
    }

    private void startReader(){
        Thread tw = new Thread(new ReaderThread());
        tw.start();
    }


    private void handleUpdate(Protobufs.Response updateResponse){
        if(updateResponse.getType() == Protobufs.Response.Type.TEST_ADDED){
            Test test = ProtoUtils.getTest(updateResponse);
            client.refreshAll(test);
        }

    }

    private class ReaderThread implements Runnable{
        public void run(){
            while(!finished){
                try{
                    Protobufs.Response response = Protobufs.Response.parseDelimitedFrom(input);
                    System.out.println("response received " + response);

                    if(isUpdateResponse(response.getType())){
                        handleUpdate(response);
                    }else{
                        try{
                            qresponses.put(response);
                        }catch(InterruptedException e){
                            e.printStackTrace();
                        }
                    }
                }catch(IOException e){
                    System.out.println("Reading error " + e);
                }
            }
        }

    }

    private boolean isUpdateResponse(Protobufs.Response.Type type){
        switch(type){
            case TEST_ADDED:
                return true;
        }
        return false;
    }
}

File: MPP/MPP Proiect/CSharp/protobuf/ProtoProxy.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Threading;
using Google.Protobuf;
using Java;
using services;

namespace protobuf{
    public class ProtoProxy: IServices{
        private readonly string host;
        private readonly int port;
        private IObserver refereeObserver;
        private NetworkStream stream;
        private IFormatter formatter;
        private TcpClient connection;
        private readonly Queue<Response> responses;
        private volatile bool finished;
        private EventWaitHandle waitHandle;

        public ProtoProxy(string host, int port){
            this.host = host;
            this.port = port;
            responses = new Queue<Response>();
        }

        private void CloseConnection(){
            finished = true;
            try{
                stream.Close();
                connection.Close();
                waitHandle.Close();
                refereeObserver = null;
                Console.WriteLine("Connection closed!");
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }
        }

        private void SendRequest(Request request){
            try{
                request.WriteDelimitedTo(stream);
                stream.Flush();
            } catch(Exception e){
                throw new MyException("Error sending object " + e);
            }
        }

        private Response ReadResponse(){
            Response response = null;
            try{
                waitHandle.WaitOne();
                lock(responses){
                    response = responses.Dequeue();
                }
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }

            return response;
        }

        private void InitializeConnection(){
            try{
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                waitHandle = new AutoResetEvent(false);
                StartReader();
            } catch(Exception e){
                Console.WriteLine(e.StackTrace);
            }
        }

        private void StartReader(){
            var tw = new Thread(Run);
            tw.Start();
        }

        private void HandleUpdate(Response response){
            if(response.Type != Response.Types.Type.TestAdded) return;
            var test = ProtoUtils.GetTest(response);
            Console.WriteLine("Test added " + test);
            try{
                refereeObserver.RefreshAll(test);
            } catch(MyException e){
                Console.WriteLine("Error handle update: " + e);
            }
        }

        private static bool IsUpdate(Response response){
            return response.Type == Response.Types.Type.TestAdded;
        }

        protected virtual void Run(){
            while(!finished){
                try{
                    var response = Response.Parser.ParseDelimitedFrom(stream);
                    Console.WriteLine("response received " + response);
                    if(IsUpdate((Response)response)){
                        HandleUpdate((Response)response);
                    } else{
                        lock(responses){
                            responses.Enqueue((Response)response);
                        }

                        waitHandle.Set();
                    }
                } catch(Exception e){
                    Console.WriteLine("Reading error " + e);
                }
            }
        }

        public model.Referee Login(model.Referee referee, IObserver client){
            InitializeConnection();
            Console.WriteLine("Login request ...");
            var req = ProtoUtils.CreateLoginRequest(referee);
            SendRequest(req);
            var response = ReadResponse();
            switch(response.Type){
                case Response.Types.Type.Ok:
                    refereeObserver = client;
                    Console.WriteLine("Logged in");
                    return ProtoUtils.GetReferee(response);
                case Response.Types.Type.Error:{
                    var err = ProtoUtils.GetError(response);
                    Console.WriteLine("Error logging in" + err);
                    CloseConnection();
                    throw new MyException(err);
                }
                default:
                    return null;
            }
        }

        public IEnumerable<model.Participant> GetAllParticipantsSortedByName(){
            var req = ProtoUtils.CreateAllParticipantsRequest();
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == Response.Types.Type.Error)
            {
                var err = ProtoUtils.GetError(response);
                Console.WriteLine("Error getting shows" + err);
                throw new MyException(err);
            }
            Console.WriteLine("Got shows");
            return ProtoUtils.GetParticipants(response);
        }

        public void AddTest(model.Test test){
            var req = ProtoUtils.CreateAddTestRequest(test);
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == Response.Types.Type.Error)
            {
                var err = ProtoUtils.GetError(response);
                Console.WriteLine("Error adding test" + err);
                throw new MyException(err);
            }
            Console.WriteLine("Test added");
        }

        public IEnumerable<model.Participant> GetAllActiveParticipants(int refereeId){
            var req = ProtoUtils.CreateAllActiveParticipantsRequest(refereeId);
            SendRequest(req);
            var response = ReadResponse();
            if (response.Type == Response.Types.Type.Error)
            {
                var err = ProtoUtils.GetError(response);
                Console.WriteLine("Error getting active participants" + err);
                throw new MyException(err);
            }
            Console.WriteLine("Got active participants");
            return ProtoUtils.GetParticipants(response);
        }

        public void Logout(model.Referee referee){
            var req = ProtoUtils.CreateLogoutRequest(referee);
            SendRequest(req);
            var response = ReadResponse();
            CloseConnection();
            if (response.Type == Response.Types.Type.Error)
            {
                var err = ProtoUtils.GetError(response);
                Console.WriteLine("Error logging out" + err);
                throw new MyException(err);
            }
            Console.WriteLine("Logged out");
        }
    }
}